File: 2025-11/2025-11-01/signature_validation.py
import re

def verify(message, key, signature):
    value_of_lowercase_a = 1
    value_of_uppercase_a = 27

    message_and_key_values = [
        ord(letter) - ord('a') + value_of_lowercase_a if re.match("[a-z]", letter) 
        else ord(letter) - ord('A') + value_of_uppercase_a 
        
        for letter in re.sub("[^a-zA-Z]", "", message + key)
    ]
    
    return sum(message_and_key_values) == signature

print(verify("foo", "bar", 57))
print(verify("foo", "bar", 54))
print(verify("freeCodeCamp", "Rocks", 238))
print(verify("Is this valid?", "No", 210))
print(verify("Is this valid?", "Yes", 233))
print(verify("Check out the freeCodeCamp podcast,", "in the mobile app", 514))

# --- TEST SUITE ---

tests_text = '''
1. verify("foo", "bar", 57) should return True.
2. verify("foo", "bar", 54) should return False.
3. verify("freeCodeCamp", "Rocks", 238) should return True.
4. verify("Is this valid?", "No", 210) should return False.
5. verify("Is this valid?", "Yes", 233) should return True.
6. verify("Check out the freeCodeCamp podcast,", "in the mobile app", 514) should return True.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 2025-11/2025-11-01/signatureValidation.js
function verify(message, key, signature) {
  const valueOfLowercaseA = 1;
  const valueOfUppercaseA = 27;

  return (message + key)
    .replace(/[^a-zA-Z]/g, "")
    .split("")
    .reduce((sum, letter) => sum += /[a-z]/.exec(letter) 
      ? letter.charCodeAt(0) - 'a'.charCodeAt(0) + valueOfLowercaseA 
      : letter.charCodeAt(0) - 'A'.charCodeAt(0) + valueOfUppercaseA
    , 0) === signature;
}

console.log(verify("foo", "bar", 57));
console.log(verify("foo", "bar", 54));
console.log(verify("freeCodeCamp", "Rocks", 238));
console.log(verify("Is this valid?", "No", 210));
console.log(verify("Is this valid?", "Yes", 233));
console.log(verify("Check out the freeCodeCamp podcast,", "in the mobile app", 514));

// --- TEST SUITE ---

const testsText = `
1. verify("foo", "bar", 57) should return true.
2. verify("foo", "bar", 54) should return false.
3. verify("freeCodeCamp", "Rocks", 238) should return true.
4. verify("Is this valid?", "No", 210) should return false.
5. verify("Is this valid?", "Yes", 233) should return true.
6. verify("Check out the freeCodeCamp podcast,", "in the mobile app", 514) should return true.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);